src/storage/mvcc/reader.rs

```diff
@@ -401,9 +413,140 @@ impl MvccReader {
                     },
                 }
             };
-            if let Some(v) = self.get(&key, ts)? {
+
+            if let Some(v) = self.reverse_get_impl(&key, ts)? {
                 return Ok(Some((key, v)));
             }
+
+            // reverse_get_impl may call write_cursor's prev.
+            if !self.write_cursor.as_ref().unwrap().valid() {
+                write_valid = false;
+            }
+        }
+    }
+
+    fn reverse_get_impl(&mut self, user_key: &Key, ts: u64) -> Result<Option<Value>> {
+        assert!(self.lock_cursor.is_some());
+        assert!(self.write_cursor.is_some());
+
+        // Check lock.
+        match self.isolation_level {
+            IsolationLevel::SI => {
+                let l_cur = self.lock_cursor.as_ref().unwrap();
+                if l_cur.valid() && l_cur.key() == user_key.encoded().as_slice() {
+                    self.statistics.lock.processed += 1;
+                    self.check_lock_impl(user_key, ts, Lock::parse(l_cur.value())?)?;
+                }
+            }
+            IsolationLevel::RC => {}
+        }
+
+        // Get value for this user key.
+        // At first, we use several prev to try to get the latest version.
+        let mut lastest_version = (None /* start_ts */, None /* short value */);
+        let mut last_handled_key: Option<Vec<u8>> = None;
+        for _ in 0..REVERSE_SEEK_BOUND {
+            if !self.write_cursor.as_mut().unwrap().valid() {
+                return self.get_value(user_key, lastest_version.0, lastest_version.1);
+            }
+
+            let mut write = {
+                let (commit_ts, key) = {
+                    let w_cur = self.write_cursor.as_ref().unwrap();
+                    last_handled_key = Some(w_cur.key().to_vec());
+                    let w_key = Key::from_encoded(w_cur.key().to_vec());
+                    (w_key.decode_ts()?, w_key.truncate_ts()?)
+                };
+
+                // reach neighbour user key or can't see this version.
+                if ts < commit_ts || &key != user_key {
+                    assert!(&key <= user_key);
+                    return self.get_value(user_key, lastest_version.0, lastest_version.1);
+                }
+                self.statistics.write.processed += 1;
```

Comment:
processed means we have found the key, is it proper here?

Reply:
I think processed means we processed this key?